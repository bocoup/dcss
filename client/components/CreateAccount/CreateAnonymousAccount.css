.caa__header-centered {
    text-align: center;
}
