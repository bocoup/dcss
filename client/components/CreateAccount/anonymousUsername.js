const animals = 'aardvark akita albatross alligator alpaca angelfish ant anteater antelope ape armadillo baboon badger bandicoot barnacle barracuda bat beagle bear beaver bee beetle bird bison bloodhound boar bobcat bonobo buffalo bulldog bullfrog butterfly caiman camel capybara caracal caribou cassowary cat caterpillar catfish cattle centipede chameleon chamois cheetah chicken chihuahua chimpanzee chinchilla chinook chipmunk clam cobra collie coral cougar coyote crab crane crocodile crow cuttlefish dachshund dalmatian deer dingo dinosaur dodo dogfish dolphin donkey dormouse dotterel dove dragonfly duck eagle eel elephant elk emu falcon ferret finch fish flamingo flounder fox frog gazelle gecko gerbil gibbon giraffe gnu goat goldfinch goldfish goose gopher gorilla goshawk grasshopper greyhound grouse gull guppy hamster hare harrier hawk hedgehog heron herring himalayan hippopotamus hornet horse human hummingbird hyena ibis iguana impala indri insect jackal jaguar bluejay jellyfish kangaroo kelpie kingfisher kiwi koala dragon labradoodle ladybird lapwing lark lemming lemur leopard liger lion lionfish lizard llama lobster locust loris lynx lyrebird macaw magpie mallard manatee mandrill mastiff meerkat mink mole mongoose monkey moorhen moose mosquito moth mouse mule narwhal newt nightingale ocelot octopus okapi opossum orangutan oryx ostrich otter owl ox oyster pademelon panther parrot partridge peacock peafowl pelican penguin persian pheasant pig pigeon piranha platypus pointer pony poodle porcupine porpoise possum prairie dog prawn puffin pug puma quail rabbit raccoon ragdoll rail ram rat rattlesnake raven panda reindeer rhinoceros robin rook rottweiler salamander salmon sand-dollar sandpiper sardine scorpion sea-lion seahorse seal shark sheep shrew shrimp skunk sloth snail snake snowshoe sparrow spider sponge squid squirrel starfish starling stingray stork swallow swan tang tapir tarsier termite tetra tiger toad tortoise toucan trout turkey turtle umbrellabird urchin viper vulture wallaby walrus warthog wasp weasel whale whippet wildebeest wolf wolverine wombat woodpecker wren xeme xolo yak yellowhammer zebra zebu zonkey zorse'.split(
  ' '
);

const adjectives = 'accountable adaptable adept adventurous affable agreeable alert alluring amazing ambitious amiable amusing attentive awesome boundless brave bright calm capable charming chatty cheerful clever coherent comfortable confident conscientious considerate consistent convivial cooperative courageous creative credible cultured dashing dazzling debonair decisive decorous delightful dependable detailed determined diligent diplomatic discerning discreet dynamic eager eclectic efficient elated eminent enchanting encouraging endurable energetic enterprising entertaining enthusiastic epic excellent excited exclusive exuberant fabulous fair fair-minded faithful fantastic favorable fearless fine flourished flowing focused forceful forgiving fortuitous frank free free-spirited friendly fulfilled fun fun-loving funny generous genial genius gentle genuine giving glad glorious glowing goddess good goodhealth goodness graceful gracious grateful great gregarious groundbreaking grounded happy happy-hearted hard-working hardworking harmonious healthy heartfull heartwarming heaven helpful high-spirited hilarious holy honest honorable hopeful humorous illuminated imaginative impartial incomparable incredible independent industrious ineffable ingenious innovative inscrutable insightful inspirational inspired instinctive intellectual intelligent intuitive inventive invigorated involved irresistible jazzed jolly jovial joyful joyous jubilant juicy just juvenescent kalon kind kind-hearted kissable knowingly knowledgeable kooky level likeable lively logical lovable loved lovely loving loyal lucky luxurious magical magnificent marvelous mature memorable meticulous mind-blowing mindful miracle miraculous mirthful modern modest multi-talented neat nice nirvana noble nourished nurtured obedient observant open open-hearted open-minded optimistic opulent organised organized original outstanding owning-my-power painstaking passionate patient peaceful perceptive perfect persistent philosophical pioneering placid plausible playful pleasant plucky polite positive powerful practical precious pro-active proactive productive professional propitious prosperous protective proud punctual quick-witted quiet radiant radical rational ready receptive reflective refreshed rejuvenated relaxed reliable relieved remarkable renewed reserved resilient resolute resourceful responsible rhetorical rich righteous robust romantic sacred safe satisfied secured sedate seemly selective self-accepting self-assured self-confident self-disciplined self-loving sensational sensible sensitive serene sharp shining shrewd shy silly sincere skillful smart smiling sociable sophisticated soulful spectacular splendid steadfast stellar stimulating straightforward strategic strong studious stupendous successful succinct super sustained sympathetic sympathetic* talented thankful thoughtful thrifty thrilled thriving tidy tough tranquil triumphant trusting trustworthy ultimate unassuming unbelievable unbiased understanding unique unlimited unprecedented unreal unusual upbeat uplifted valuable versatile vibrant victorious vigorous vivacious voracious warm warmhearted wealthy welcomed whole wholeheartedly willing wise witty wonderful wondrous worthy xoxo young-at-heart youthful yummy zappy zealous zestful'.split(
  ' '
);

export default function() {
  const animal = animals[Math.floor(Math.random() * animals.length)];
  const adjective = adjectives[Math.floor(Math.random() * adjectives.length)];
  return `${adjective}-${animal}`;
}
