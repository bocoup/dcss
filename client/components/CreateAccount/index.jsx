import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Button, Form, Grid } from 'semantic-ui-react';

import { logIn } from '@client/actions';

class CreateAccount extends Component {
    constructor(props) {
        super(props);
        this.state = {
            usernameInput: '',
            emailInput: '',
            passwordInput: '',
            confirmPasswordInput: '',
            createMessage: ''
        };

        this.validFormInput = this.validFormInput.bind(this);
        this.onSubmit = this.onSubmit.bind(this);
        this.onChange = this.onChange.bind(this);
    }

    validFormInput() {
        if (!this.state.passwordInput || !this.state.confirmPasswordInput) {
            this.setState({
                createMessage:
                    'Please enter and confirm your intended password.'
            });
            return false;
        }

        if (this.state.passwordInput !== this.state.confirmPasswordInput) {
            this.setState({
                createMessage:
                    'Password fields do not match. Please confirm your intended password.'
            });
            return false;
        }

        if (!this.state.usernameInput) {
            this.setState({
                createMessage: 'Please enter a username for your account.'
            });
            return false;
        }

        return true;
    }

    async onSubmit() {
        if (!this.validFormInput()) {
            return;
        }
        const data = JSON.stringify({
            username: this.state.usernameInput,
            email: this.state.emailInput,
            password: this.state.passwordInput
        });
        const createResponse = await (await fetch('/api/auth/signup', {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: data
        })).json();

        if (createResponse.error) {
            this.setState({ createMessage: createResponse.message });
            return;
        }

        if (createResponse.username) {
            this.props.logIn(createResponse.username);
            this.props.history.push('/');
        }
    }

    onChange(event) {
        if (this.state.createMessage) {
            this.setState({ createMessage: '' });
        }
        this.setState({ [`${event.target.name}Input`]: event.target.value });
    }

    render() {
        return (
            <Form className="signup__form">
                <Form.Field>
                    <label htmlFor="name">Username</label>
                    <input
                        required
                        name="username"
                        autoComplete="username"
                        onChange={this.onChange}
                        value={this.state.usernameInput}
                    />
                </Form.Field>
                <Form.Field>
                    <label htmlFor="email">Email Address</label>
                    <input
                        name="email"
                        autoComplete="email"
                        placeholder="(Optional)"
                        onChange={this.onChange}
                        value={this.state.emailInput}
                    />
                </Form.Field>
                <Form.Field>
                    <label htmlFor="password">Password</label>
                    <input
                        required
                        name="password"
                        type="password"
                        autoComplete="new-password"
                        onChange={this.onChange}
                        value={this.state.passwordInput}
                    />
                </Form.Field>
                <Form.Field>
                    <label htmlFor="confirmPassword">Confirm Password</label>
                    <input
                        required
                        name="confirmPassword"
                        type="password"
                        autoComplete="new-password"
                        onChange={this.onChange}
                        value={this.state.confirmPasswordInput}
                    />
                </Form.Field>
                <Grid columns={2}>
                    <Grid.Column>
                        <Button
                            primary
                            type="submit"
                            size="large"
                            onClick={this.onSubmit}
                        >
                            Create Account
                        </Button>
                    </Grid.Column>
                    <Grid.Column>{this.state.createMessage}</Grid.Column>
                </Grid>
            </Form>
        );
    }
}

CreateAccount.propTypes = {
    history: PropTypes.shape({
        push: PropTypes.func.isRequired
    }).isRequired,
    logIn: PropTypes.func.isRequired,
    logOut: PropTypes.func.isRequired,
    isLoggedIn: PropTypes.bool.isRequired,
    username: PropTypes.string
};

function mapStateToProps(state) {
    const { isLoggedIn, username } = state.login;
    return { isLoggedIn, username };
}

const mapDispatchToProps = {
    logIn
};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(CreateAccount);
