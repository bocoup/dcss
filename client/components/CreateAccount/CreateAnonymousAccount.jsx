import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import PropTypes from 'prop-types';
import { Button, Header, Modal, Popup } from 'semantic-ui-react';
import { logIn } from '@actions';
import Loading from '@components/Loading';
import anonymousUsername from './anonymousUsername';
import './CreateAnonymousAccount.css';

class CreateAnonymousAccount extends Component {
  constructor(props) {
    super(props);

    const from = this.props.location.state && this.props.location.state.from;

    this.state = {
      from,
      username: ''
    };

    this.onClick = this.onClick.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
  }

  componentDidMount() {
    this.generateUnusedAnonymousUsername();
  }

  async generateUnusedAnonymousUsername() {
    const username = anonymousUsername();
    // TODO: Move to own async action
    const { status } = await (await fetch(
      `/api/auth/signup/usernames/${username}/exists`
    )).json();

    if (status === 409) {
      return await this.generateUnusedAnonymousUsername();
    }

    if (username) {
      this.setState({ username });
    }
  }

  async onSubmit() {
    const { from, username } = this.state;
    const body = JSON.stringify({
      username
    });

    const { error, message } = await (await fetch('/api/auth/signup', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body
    })).json();

    if (error) {
      this.setState({ message });
    } else {
      // Step outside of react to force a real reload
      // after signup and session create
      location.href = from ? `${from.pathname}${from.search}` : '/';
    }
  }

  onClick(event, { name }) {
    if (name === 'no') {
      this.generateUnusedAnonymousUsername();
    }
    if (name === 'nvm') {
      this.props.history.push('/login/create-account');
    }
  }

  render() {
    const { username } = this.state;
    const { onClick, onSubmit } = this;
    const header = username
      ? 'Do you like this anonymous user name?'
      : 'Find an anonymous user name for you.';
    return (
      <Modal open size="small">
        <Header icon="user outline" content={header} />
        <Modal.Content>
          {username ? (
            <Header as="h1" className="createanonymousaccount__header-centered">
              <code>{username}</code>
            </Header>
          ) : (
            <Loading />
          )}
        </Modal.Content>
        <Modal.Actions style={{ height: '75px' }}>
          <Button.Group fluid size="large">
            <Popup
              content="Go back"
              position="bottom center"
              trigger={
                <Button
                  content="Nevermind, go back"
                  name="nvm"
                  onClick={onClick}
                />
              }
            />
            <Button.Or />
            <Popup
              content="Show me another available user name."
              position="bottom center"
              trigger={
                <Button
                  color="orange"
                  content="No, see another"
                  name="no"
                  onClick={onClick}
                />
              }
            />
            <Button.Or />
            <Popup
              content="Select this user name to continue."
              position="bottom center"
              trigger={
                <Button
                  color="green"
                  content="Yes, continue"
                  name="yes"
                  onClick={onSubmit}
                />
              }
            />
          </Button.Group>
        </Modal.Actions>
      </Modal>
    );
  }
}

CreateAnonymousAccount.propTypes = {
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }).isRequired,
  location: PropTypes.object,
  logIn: PropTypes.func.isRequired,
  isLoggedIn: PropTypes.bool.isRequired,
  username: PropTypes.string
};

const mapStateToProps = state => {
  const { isLoggedIn, username } = state.login;
  return { isLoggedIn, username };
};

const mapDispatchToProps = {
  logIn
};

export default withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(CreateAnonymousAccount)
);
