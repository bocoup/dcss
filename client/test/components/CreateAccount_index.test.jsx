import {
  fetchImplementation,
  mounter,
  reduxer,
  snapshot,
  state
} from '../bootstrap';
import { unmountComponentAtNode } from 'react-dom';
import { cleanup, fireEvent, render } from '@testing-library/react';
import CreateAccount from '../../components/CreateAccount/index.jsx';

beforeAll(() => {
  (window || global).fetch = jest.fn();
});

let container = null;
beforeEach(() => {
  container = document.createElement('div');
  container.setAttribute('id', 'root');
  document.body.appendChild(container);

  fetchImplementation(fetch);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

const sharedProps = {
  history: {
    push() {}
  }
};

test('CreateAccount', () => {
  expect(CreateAccount).toBeDefined();
});

test('Snapshot 1', () => {
  const props = {
    ...sharedProps
  };
  const mounted = mounter(reduxer(CreateAccount, props, state))();
  expect(snapshot(mounted)).toMatchSnapshot();
});

/*{INJECTION}*/
