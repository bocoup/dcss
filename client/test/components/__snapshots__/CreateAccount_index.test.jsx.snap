// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1 1`] = `"<div class=\\"ui page modals dimmer transition visible active\\" style=\\"display: flex;\\"><div role=\\"dialog\\" aria-modal=\\"true\\" aria-labelledby=\\"create-a-new-user-account\\" class=\\"ui small modal transition visible active\\"><i aria-hidden=\\"true\\" class=\\"close icon\\"></i><div id=\\"create-a-new-user-account\\" class=\\"ui header\\"><i aria-hidden=\\"true\\" class=\\"user outline icon\\"></i><div class=\\"content\\">Create a new user account</div></div><div class=\\"content\\"><form class=\\"ui form\\"><div class=\\"field\\"><div class=\\"required field\\"><label>Username:</label><div class=\\"ui input\\"><input name=\\"username\\" autocomplete=\\"off\\" required=\\"\\" type=\\"text\\" value=\\"\\"></div></div></div><div class=\\"field\\"><div class=\\"field\\"><label>Email address:</label><div class=\\"ui input\\"><input name=\\"email\\" autocomplete=\\"email\\" placeholder=\\"(Optional)\\" type=\\"text\\" value=\\"\\"></div></div></div><div class=\\"field\\"><div class=\\"required field\\"><label>Password:</label><div class=\\"ui input\\"><input name=\\"password\\" autocomplete=\\"new-password\\" required=\\"\\" type=\\"password\\" value=\\"\\"></div></div></div><div class=\\"field\\"><div class=\\"required field\\"><label>Confirm password:</label><div class=\\"ui input\\"><input name=\\"confirmPassword\\" autocomplete=\\"new-password\\" required=\\"\\" type=\\"password\\" value=\\"\\"></div></div></div></form></div><div class=\\"actions\\"><div class=\\"ui grid\\"><div class=\\"row\\"><div class=\\"column\\"><div class=\\"ui fluid buttons\\"><button type=\\"submit\\" class=\\"ui large primary button\\">Submit</button><div class=\\"or\\"></div><button class=\\"ui large button\\">Cancel</button></div><div style=\\"text-align: left;\\" class=\\"ui red floating hidden message\\"></div></div></div></div></div></div></div>"`;
